File: 21553 - Sai Nandan D/Week 6/215_Kth Largest Element in an Array.py
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:        
        n = len(nums)
        
        def partition(lo, hi):
            r_i = randint(lo, hi)
            nums[r_i], nums[hi] = nums[hi], nums[r_i]
            pivot = nums[hi]
            l = lo-1
            for r in range(lo, hi):
                if nums[r] >= pivot:
                    l += 1
                    nums[l], nums[r] = nums[r], nums[l]
            nums[l+1], nums[hi] = nums[hi], nums[l+1]
            return l+1
            
    
        def kth_largest(lo, hi, k):
            if lo == hi: return nums[lo]
            p = partition(lo, hi)
            count = p - lo + 1
            if count > k: return kth_largest(lo, p-1, k)
            if count < k: return kth_largest(p+1, hi, k-count)            
            return nums[p]
            
        return kth_largest(0, n-1, k)

File: 21561 - Suvam Tamang/Week 6/215_Kth Largest Element in an Array.java
class Solution {
        
    public int findKthLargest(int[] nums, int k) {
        if(nums.length == 1) 
            return nums[0];
        int left = 0;
        int right = nums.length - 1;
        while(left <= right){
            int pivotPos = partition(nums, left, right);
            int count = pivotPos - left + 1;
            if(count < k){
                k = k - count;
                left = pivotPos + 1;
            }else if(count > k){
                right = pivotPos - 1;
            }else{
                return nums[pivotPos];
            }
        }
        return 0;
    }
    
    public int partition(int[] nums,int left,int right) {
        int pivot = nums[right];
        int leftBound = left;
        int rightBound = right - 1;
        
        while(leftBound <= rightBound) {
            if(nums[leftBound] >= pivot)
                leftBound++;
            else if(nums[rightBound] < pivot)
                rightBound--;
            else 
                swap(nums, leftBound++, rightBound--);
        }
        swap(nums, leftBound, right);
        return leftBound;
    }
    
    public void swap(int[] arr,int l,int p) {
        int temp = arr[l];
        arr[l] = arr[p];
        arr[p] = temp;
    }
}
